### PR TITLE
feat: [CDS-51189]: Terraform run step delegate side logic

### PR DESCRIPTION
**Pull request by alebuh**
## Describe your changes
Introducing delegate side changes for RUN_REFRESH_STATE, RUN_PLAN_ONLY,  RUN_PLAN, RUN_PLAN_AND_APPLY, RUN_PLAN_AND_DESTROY, RUN_APPLY terraform cloud task types.
Also fixing, and changing a few things on the Manager's side 
The Exception handling is poor, but there is a separate jira ticket where this will be covered.
TERRAFORM_CLOUD_TASK_NG is unsupported and this logic won't be used until it's tested in detail 
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43578)

### CODE REVIEW

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 22 of 24 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43578)* status: 24 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/terraformcloud/TerraformCloudRunStepPlanCreator.java` line 52 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpqglr8oNVJKMTWUvx:-NNpqglr8oNVJKMTWUvy:b-dcwab9) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/125-cd-nextgen/src/main/java/io/harness/cdng/creator/plan/steps/terraformcloud/TerraformCloudRunStepPlanCreator.java#L52)):*
> ```Java
>     TerraformCloudRunStepParameters terraformCloudRunStepParameters =
>         (TerraformCloudRunStepParameters) ((StepElementParameters) stepParameters).getSpec();
>     terraformCloudRunStepParameters.setStepFqn(stepFqn);
> ```

We should avoid using stepFqn. I do believe I've already discussed with Tudor how this can be solved, that you can instead of step fqn and output name rely on the terraform plan details entity id. This is a good opportunity to test this logic on TFE first before refactoring it for Terraform 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/RunRequestCreator.java` line 92 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNprVgoEzTrJIS9rwfw:-NNprVgoEzTrJIS9rwfx:b-u7kt0j) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/RunRequestCreator.java#L92)):*
> ```Java
>               .map(entry -> Variable.builder().key(entry.getKey()).value(entry.getValue()).build())
>               .collect(Collectors.toList())
>         : null;
> ```

We should return emptyList instead
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java` line 73 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpoHew2WaxTtWrITQw:-NNpoHew2WaxTtWrITQx:bcd3040) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java#L73)):*
> ```Java
> @Slf4j
> @Singleton
> public class TerraformCloudTaskHelper {
> ```

I see the issue with this class as is not actually a helper method but all the logic is basically a task handler. In my opinion we should avoid this code structure, every method in the helper should do 1 thing and is the task handler who should orchestrate everything. Otherwise this code may end up in a complex logic where is going to be hard to understand and make any changes as basically we're hiding the complexity of the task handler behind the helper method. 
I'll mark this as suggestion, but please take this in consideration while feature is not yet released
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java` line 191 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpmz6E-wPFXWJQWuQP:-NNpmz6E-wPFXWJQWuQQ:b-3gva82) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java#L191)):*
> ```Java
>       logCallback.saveExecutionLog("Refresh state FAILED", LogLevel.ERROR, CommandExecutionStatus.FAILURE);
>       log.error(format("Refresh State run failed: %s", e.getMessage()));
>       return TerraformCloudRunTaskResponse.builder().commandExecutionStatus(CommandExecutionStatus.FAILURE).build();
> ```

We should avoid using commandExecutionStatus. Every exception should be thrown and handled with exception handler if it is the case, everything else that doesn't throw an exception should be handled as a successful case
Your method interface should not directly return TerraformCloudRunTaskResponse. If, as in this case, there is no additional data that method returns then it should return nothing (void). It's the task handler who should handle this. 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java` line 205 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpmYNC08WGyWVIdjYK:-NNpmYNC08WGyWVIdjYL:b1cfnv3) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java#L205)):*
> ```Java
>       if (taskParameters.isExportJsonTfPlan()) {
>         String planId = getRelationshipId(runData, "plan");
>         String jsonPlan = terraformCloudClient.getPlanJsonOutput(url, token, planId);
> ```

I would suggest to rewrite this API to use stream based response. This is not going to scale well if we're reading initially the json plan in memory. As a json plan can be 100mb large in parallel executions this will end up in OOM
Marking this as suggestion but please take in consideration for further work
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java` line 229 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNppWOB9zw5B7JbXmHv:-NNppWOB9zw5B7JbXmHw:b-er1jia) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java#L229)):*
> ```Java
>   public TerraformCloudRunTaskResponse autoApply(
>       TerraformCloudConfig terraformCloudConfig, TerraformCloudTaskParams taskParameters, LogCallback logCallback) {
>     String url = getUrlFromTerraformCloudConfig(terraformCloudConfig);
> ```

Shouldn't this be part of client logic? 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java` line 257 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpnlQ79nwMwIQTe-A3:-NNpnlQ79nwMwIQTe-A4:bvvvsev) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java#L257)):*
> ```Java
>       String status = terraformCloudClient.getRun(url, token, runId).getData().getAttributes().getStatus();
> 
>       if (status.equals("policy_checked")) {
> ```

This should be a constant. I would suggest that the client API should return an enum instead of the string 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java` line 377 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpq22G1yRZCv7ylvIB:-NNpq22G1yRZCv7ylvIC:bmpj2iz) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/930-delegate-tasks/src/main/java/io/harness/delegate/task/terraformcloud/TerraformCloudTaskHelper.java#L377)):*
> ```Java
> 
>   private String getRelationshipId(Data data, String relationshipName) {
>     return data.getRelationships()
> ```

We need more guards here, it can end up in NPE
___
*[`960-api-services/src/main/java/io/harness/terraformcloud/model/Attributes.java` line 26 at r1](https://reviewable.io/reviews/harness/harness-core/43578#-NNpqUzr2bqTKicyBeGO:-NNpqUzr2bqTKicyBeGP:b-cfmeop) ([raw file](https://github.com/harness/harness-core/blob/8d792f8c38780da41eec2eca1c8cd307f775ad56/960-api-services/src/main/java/io/harness/terraformcloud/model/Attributes.java#L26)):*
> ```Java
> public class Attributes {
>   private String message;
>   private String status;
> ```

This should be an enum

**Review by abosii-harness — APPROVED** [review]
> Reviewed 8 of 8 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43578)* status: 28 file(s) to review. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/functor/TerraformPlanJsonFunctor.java` line 86 at r3](https://reviewable.io/reviews/harness/harness-core/43578#-NOJ6iZS7XufCvMbF9Xx:-NOJ6iZTDibuXyTpAS-U:bh8sl7n) ([raw file](https://github.com/harness/harness-core/blob/0045fc6c622bdc4c74a5b20f577bd24eecbb0842/125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/functor/TerraformPlanJsonFunctor.java#L86)):*
> ```Java
> 
>     return terraformPlanExectionDetailsService
>         .listAllPipelineTFPlanExecutionDetails(TFPlanExecutionDetailsKey.builder()
> ```

nit: can we just find it by id? It should be faster
___
*[`960-api-services/src/main/java/io/harness/terraformcloud/model/RunStatus.java` line 10 at r3](https://reviewable.io/reviews/harness/harness-core/43578#-NOJ8wac3GyRH4PQreDB:-NOJ8wac3GyRH4PQreDC:ba87q46) ([raw file](https://github.com/harness/harness-core/blob/0045fc6c622bdc4c74a5b20f577bd24eecbb0842/960-api-services/src/main/java/io/harness/terraformcloud/model/RunStatus.java#L10)):*
> ```Java
> package io.harness.terraformcloud.model;
> 
> public enum RunStatus {
> ```

nit: Could be worth to add @JsonEnumDefaultValue
You may need to update retrofit factory for jackson
.addConverterFactory(JacksonConverterFactory.create(new ObjectMapper().enable(READ_UNKNOWN_ENUM_VALUES_USING_DEFAULT_VALUE)))
___
*[`960-api-services/src/main/java/io/harness/terraformcloud/model/RunStatus.java` line 11 at r3](https://reviewable.io/reviews/harness/harness-core/43578#-NOJ7EfE1b39jsBeoTzP:-NOJ7EfE1b39jsBeoTzQ:b86vu1d) ([raw file](https://github.com/harness/harness-core/blob/0045fc6c622bdc4c74a5b20f577bd24eecbb0842/960-api-services/src/main/java/io/harness/terraformcloud/model/RunStatus.java#L11)):*
> ```Java
> 
> public enum RunStatus {
>   pending,
> ```

nit: You can use @JsonProperty here and keep the java compatible case for enums

File: 960-api-services/src/main/java/io/harness/terraformcloud/model/Attributes.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  private RunStatus status;
+  private String source;
+  @JsonProperty("terraform-version") private String terraformVersion;
+  @JsonProperty("plan_and_apply") private boolean planAndApply;
+  @JsonProperty("plan_only") private boolean planOnly;
+  @JsonProperty("refresh_only") private boolean refreshOnly;
+  @JsonProperty("is-destroy") private boolean isDestroy;
+  @JsonProperty("empty_apply") private boolean emptyApply;
+  @JsonProperty("auto-apply") private boolean autoApply;
+  @JsonProperty("target-addrs") private List<String> targets;
```
Comment (vlica-harness) [inline 960-api-services/src/main/java/io/harness/terraformcloud/model/Attributes.java:35]: @JsonProperty("target-addrs") private List<String> targets;
Is it a typo here target-addrs?

**Review by vlica-harness — APPROVED** [review]

**Review by vlica-harness — APPROVED** [review]
> Reviewed 10 of 24 files at r1, 8 of 8 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43578)* status: :shipit: complete! ✓

File: 960-api-services/src/main/java/io/harness/terraformcloud/model/Attributes.java (hunk 1 above)
Comment (alebuh) [reply 960-api-services/src/main/java/io/harness/terraformcloud/model/Attributes.java:35]: It's not a typo (at least not on our end), that's how Terraform Enterprise API is defined

**Review by vlica-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43578)* status: :shipit: complete! ✓

**Pull request merged**